Initial prototype of aws-ecs-1 variant

samuelkarp opened the pull request:
**Issue number:**
https://github.com/bottlerocket-os/bottlerocket/issues/815


**Description of changes:**
This pull request introduces a new variant of Bottlerocket called aws-ecs-1.  The new variant includes the ECS agent and Docker to run containers.

The commits here represent the skeleton for packaging the ECS agent, defining the ECS model in the Bottlerocket API, rendering appropriate config, and making the new variant.  I'd like to merge this (after review, of course) and then continue iterating on the variant in subsequent PRs.

**Testing done:**
Launched an AMI from this variant, observed that it connected to a non-default cluster (configured with the new `settings.ecs.cluster` setting) in ECS, and ran a basic task with minimal features (busybox:latest running sleep).

Set the new `settings.ecs.instance-attributes` setting through TOML-formatted user-data.  Observed that the attributes are registered properly, along with the default `bottlerocket.variant` and `bottlerocket.version` attributes.

Toggled the new `settings.ecs.privileged-containers-enabled` setting through the Bottlerocket API.  Observed that the config file was properly regenerated, the agent restarted, and the capabilities updated in the ECS API.

Validated that the pause container image is automatically loaded and able to run.

If you wish to test this variant, you can use this user-data:

```toml
[settings.ecs]
cluster = "bottlerocket"

[settings.ecs.instance-attributes]
hello = "world"

[settings.host-containers.admin]
enabled = true
```

Note: I have not yet updated the top-level README as the variant is not really ready for use yet.  This means that the settings are currently undocumented.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

samuelkarp committed:
variants: Remove reference to top-level list

samuelkarp committed:
variants: create aws-ecs-1 variant

samuelkarp committed:
models: add initial model for aws-ecs-1

samuelkarp committed:
docker-engine: use cgroupfs cgroup driver

The systemd cgroup driver imposes additional restrictions on the name of
cgroup parents, causing workflows that depend on non-conforming names to
fail.

bcressey commented on packages/docker-engine/daemon.json:
```diff
@@ -4,7 +4,6 @@
   "live-restore": true,
   "max-concurrent-downloads": 10,
   "storage-driver": "overlay2",
-  "exec-opts": ["native.cgroupdriver=systemd"],
```
Switching this to `"native.cgroupdriver=cgroupfs"` would help reinforce that this is an intentional choice rather than an omission.

samuelkarp replied:
Do we want to intentionally *choose* cgroupfs, or leave it to Docker to determine the default?  My inclination was to allow Docker to pick its default, but I'm happy to change it if you think we should.

jahkeup replied:
I'd prefer we be as specific as possible for the variant - the configuration chosen should reflect *exactly* the intended runtime configuration. This also makes debugging later much much simpler to identify choices and track down these sort of conversations.

samuelkarp replied:
Done.

bcressey commented on packages/ecs-init/ecs-init.spec:
```diff
@@ -0,0 +1,86 @@
+%global agent_image %{S:2}
+
+%global _cachedir %{_localstatedir}/cache
+%global no_exec_perm 644
```
nit: this is a bit more fine-grained than is usual for macros. Repetition for this is tolerated at the command level since `0644` is widely recognized and the macro tends to obscure rather than clarify.

samuelkarp replied:
Done.

bcressey commented on packages/ecs-init/ecs-init.spec:
```diff
@@ -0,0 +1,86 @@
+%license LICENSE NOTICE THIRD-PARTY
+%{_cross_libexecdir}/amazon-ecs-init
+%config(noreplace) %ghost %{_cross_sysconfdir}/ecs/ecs.config
+%config(noreplace) %ghost %{_cross_sysconfdir}/ecs/ecs.config.json
```
Don't need to declare `%ghost` since we delete the RPM database after installing the packages.

samuelkarp replied:
Done.

samuelkarp commented:
(Note: CI failures are due to new package sources not being available in the cache yet.)

jahkeup commented on packages/ecs-init/0001-bottlerocket-remove-interaction-with-Nvidia-lib.patch:
```diff
@@ -0,0 +1,333 @@
+From 98f7025e054be506d431501422a74ba135fe6644 Mon Sep 17 00:00:00 2001
```
If you haven't done so already, please open an issue with the debugging that's already been done with relevant snippets of conversation and logged output.

samuelkarp replied:
There's now a comment in the spec that points to what I believe is the cause.  Per our offline discussion, however, I'm planning to move away from using ecs-init.  I may need a similar patch for the ECS agent's source code and can either annotate directly in the patch/spec, or can open an issue at that time.

jahkeup commented on packages/ecs-init/0002-docker-do-not-bind-mount-missing-plugin-dirs.patch:
```diff
@@ -0,0 +1,35 @@
+From ecfb953b6b93c6e497303a7f7d32950d5d9f28f6 Mon Sep 17 00:00:00 2001
```
This seems like an "agility" improvement that could/should be submitted to upstream.

Do the downstream code paths handle the missing path as well? Could this cause issues for the Agent process?

samuelkarp replied:
The agent runs successfully without this directory mounted in.  This may be a candidate for upstream inclusion.

samuelkarp replied:
This is no longer applicable as I have moved away from using ecs-init.

jahkeup commented on packages/ecs-init/README.md:
```diff
@@ -0,0 +1,24 @@
+  create any directory that does not exist, and `/usr/lib/docker/plugins` is
+  such a directory that does not exist.  However, Bottlerocket's filesystem is
+  mostly read-only, and ecs-init fails to create this directory.  For now,
+  bind-mounting is skipped when a directory does not exist, but this will break
```
I'm not familiar with how plugins are used, what functionality would this limit?

samuelkarp replied:
Docker plugins can be installed into those directories, see [here](https://docs.docker.com/engine/extend/plugin_api/#plugin-discovery).  The agent scans these directories to discover unmanaged plugins.  However, it does look like "managed" plugins (installed via the "docker plugin install" command) are typically installed into `/run/docker/plugins` which does work on Bottlerocket.  It's probably reasonable for us to only support managed plugins on Bottlerocket, or unmanaged plugins built into Bottlerocket itself.

jahkeup commented on variants/aws-ecs-1/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+# core
+    "release",
+# docker
+    "docker-cli",
```
Do we need to have the Docker CLI available?

samuelkarp replied:
ECS doesn't depend on it, but it's mighty handy for debugging.  I'm using it to verify that the agent works correctly, but I'd be open to removing it if we think that's the right security choice.

jahkeup replied:
I think we'd probably want to reduce that surface area. My sense is that this is something that belongs in the Admin container. I might even suggest that we set `DOCKER_HOST` to allow the admin container to connect to the socket for breakglass debugging.

samuelkarp replied:
I like that idea, though we'll have to be careful about it.  The Docker CLI in the admin container could conceivably be newer than the Docker engine on the host, leading to version mismatch issues.  (I *think* the upstream Docker project introduced version detection logic to limit the impact of this, but I'd want to test it to make sure it works the way we think it does before we depend on it.)

tjkirch replied:
Maybe track removing it in the ecs-agent package README (with the rest of the TODOs) or move those to a higher-level ECS TODO list somewhere?

samuelkarp replied:
@tjkirch Done

samuelkarp committed:
packages: define %{_cross_variant} macro

This macro can be used for conditions in variant-specific packages like
`os`.

samuelkarp committed:
ecs-settings-applier: initial implementation

samuelkarp committed:
os: build ecs-settings-applier for aws-ecs-1

ecs-settings-applier is built conditionally for aws-ecs-1.  Attempting
to compile ecs-settings-applier for another variant will fail, as the
applier depends on the definition of model::Model in the ECS variant.

samuelkarp committed:
aws-ecs-1: use ecs-settings-applier

jahkeup commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,68 @@
+        Ok(s) => s,
+        Err(e) => {
+            error!("Failed to read settings: {}", e);
+            return
```
Other settings appliers exit with a non-zero status when they encounter an error. So, you'd do something like:

```suggestion
            std::process::exit(1)
```

samuelkarp replied:
I saw that some of the other programs treat the `main` fn as a [wrapper around another fn](https://github.com/bottlerocket-os/bottlerocket/blob/7303622bce4ba7f4c1587a0abf9c4a18ffb9db8f/sources/api/schnauzer/src/main.rs#L171-L176).  I assume that's the pattern I'd want to follow?  I haven't touched snafu yet though.

jahkeup replied:
Yep, I [suggested this below](https://github.com/bottlerocket-os/bottlerocket/pull/946#discussion_r445074568) and forgot to mention it here. :facepalm: 

The wrapper would be responsible for exiting with the appropriate exit code. For what its worth: you don't *have to* use Snafu, you can just as easily use a String as your "error" type to be returned to your main: `Result<(), String>`. Though, your error type should really be one that actually implements `Error`.  

Aside from the error types, I think there's a set of pre-defined codes that are handled especially but can't find where at the moment. If you need special handling, other than pass/fail, then we might want to seek that out.

jahkeup replied:
Tackling this with `snafu` though should be straightforward. This differs from Go where errors are ad-hoc - with `snafu` (and many others like it) you end up with well defined Error types. `snafu` uses enums and its variants that are used to derive "helper" types and conversion implementations.

To get started with `snafu`, I'd suggest starting out by mimicking [`pluto`](https://github.com/bottlerocket-os/bottlerocket/blob/f1c1d2379c0f8f49b2d20090e5b6f23c38cb6d5d/sources/api/pluto/src/main.rs#L48-L69). You need a `mod error` to contain the helper types and your primary `Error` enum (called [`PlutoError` there](https://github.com/bottlerocket-os/bottlerocket/blob/f1c1d2379c0f8f49b2d20090e5b6f23c38cb6d5d/sources/api/pluto/src/main.rs#L63)). Then you can use error helper macros and trait extensions that make working with `?` much more friendly (eg: with `context()` and `ensure!()`).

jahkeup replied:
I'm still eagerly awaiting [`std::process::Termination`](https://doc.rust-lang.org/std/process/trait.Termination.html) which can be used to plumb errors all the way to main so that you can say:

```rust
fn main() -> Result<(), ProgramError> {
  // program here, with liberal use of `?`
  Ok(())
}
```

However, if you don't need to signal anything other than `std::process::exit(1)` when an error is hit - you can use the above. When the main hits the `Err` variant of `Result` it will exit `1`:

```rust
// using an error crate here to demonstrate, it doesn't have to be anyhow
// (snafu, failure, or other friends).
use anyhow::anyhow;

fn main() -> anyhow::Result<()> {
    let n: i32 = rand::random();
    if n % 2 != 0 {
        return Err(anyhow!("well, there's an error"))
    }
    Ok(())
}
```

edit: termination may not be required - added example.

samuelkarp replied:
Done!

jahkeup commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,68 @@
+    cluster: Option<String>,
+}
+
+fn main() {
```
Pulling this into a function that returns a `Result` would let you use the `?` in a handful of places here.

[You can see this in practice in `pluto`](https://github.com/bottlerocket-os/bottlerocket/blob/f1c1d2379c0f8f49b2d20090e5b6f23c38cb6d5d/sources/api/pluto/src/main.rs#L281-L289).
```suggestion
fn run() -> Result<(), ()> {
```

samuelkarp replied:
Done!

jahkeup commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,68 @@
+}
+
+/// Writes the rendered data at the proper location
+fn write_to_disk<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> Option<io::Error> {
```
The transformation into an `Option` forces callers to inspect the values rather than use conventional error handling. This could instead be massaged into a lightweight, and data-less, `Result`:

###### **this is written without a compiler telling me it is ok, ymmv**

```suggestion
fn write_to_disk<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> std::io::Result<()> {
```

Then this function body can be reduced to:

```rust
fn write_to_disk<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> std::io::Result<()> {
  if let Some(dirname) = path.as_ref().parent() {
    fs::create_dir_all(dirname)?;
  }
  fs::write(path, contents).map(|_| ())
}
```

samuelkarp replied:
Done!

samuelkarp committed:
docker-engine: allow Docker to use default bridge

The default docker0 bridge allows container networking to function in a
familiar way to developers.  The bridge is used by `docker run` when
exposing ports and used by Amazon ECS.

samuelkarp committed:
ecs-settings-applier: use snafu for errors

samuelkarp committed:
ecs-settings-applier: add --socket-path arg

samuelkarp committed:
ecs-settings-applier: add default attributes

samuelkarp committed:
models: add instance attributes for aws-ecs-1

samuelkarp committed:
ecs-settings-applier: render attributes

samuelkarp committed:
aws-ecs-1: add setting for privileged containers

samuelkarp committed:
ecs-settings-applier: render privileged containers

samuelkarp committed:
aws-ecs-1: default disable privileged containers

samuelkarp committed:
ecs-agent: new package

samuelkarp committed:
aws-ecs-1: add ecs-agent package

jahkeup commented on packages/ecs-agent/ecs.service:
```diff
@@ -0,0 +1,34 @@
+ -p tcp -m tcp --dport 80 -j REDIRECT --to-ports 51679
+ExecStopPost=-/sbin/iptables -t filter -D INPUT --dst 127.0.0.0/8 ! \
+ --src 127.0.0.0/8  -m conntrack ! --ctstate RELATED,ESTABLISHED,DNAT -j DROP
+ExecStopPost=/bin/sh -c \
```
Is the intent to replace this shutdown `sysctl` tuning with something else? We shouldn't have `/bin/sh` available at runtime in the host's rootfs.

samuelkarp replied:
Yes.  We don't have `/bin/sh` available at runtime, so this command doesn't work today.  From the temporary README.md in the same directory:

> The systemd unit contains many `ExecStartPre`/`ExecStartPost` commands, with little explanation or infrastructure.  One `ExecStopPost` command depends on `/bin/sh`, which is not available in Bottlerocket or compatible with Bottlerocket's goals.

I'm not sure what that replacement will look like yet though.

bcressey replied:
Do we need to set `net.ipv4.conf.all.route_localnet` back to the default value on stop? Might be easiest to just remove it.

samuelkarp replied:
Done

samuelkarp committed:
containerd: add config template for aws-ecs-1

iliana commented on packages/ecs-agent/ecs-agent.spec:
```diff
@@ -0,0 +1,89 @@
+  ./agent
+
+# Build the pause container
+(
+  set -x
+  cd misc/pause-container/buildPause
+  mkdir -p rootfs/usr/bin
+  # TODO determine which other C and LD flags we want
+  make BIN=rootfs/usr/bin/pause GCC=%{_cross_triple}-musl-gcc
+
+  # Construct image
+  mkdir -p image/rootfs
+  tar cvf image/rootfs/layer.tar -C rootfs .
+  DIGEST=$(sha256sum image/rootfs/layer.tar | sed -e 's/ .*//')
+  echo -n "1.0" > image/rootfs/VERSION
+  install -m 0644 %{S:2} image/config.json
+  sed -i "s/~~digest~~/${DIGEST}/" image/config.json
+  install -m 0644 %{S:3} image/manifest.json
+  install -m 0644 %{S:4} image/repositories
+  tar cvf ../../../amazon-ecs-pause.tar -C image .
+)
```
If possible it might be nice to build this in a reproducible way. I don't know if GCC's output will be reproducible, but making the timestamps + order of files in the tarball reproducible is definitely doable: https://reproducible-builds.org/docs/archives/

samuelkarp replied:
Done

samuelkarp committed:
ecs-agent: build pause image archive reproducibly

See https://reproducible-builds.org/docs/archives/

bcressey commented on packages/ecs-agent/README.md:
```diff
@@ -0,0 +1,17 @@
+* Logging is currently set to `debug` to assist with development.
+* The systemd unit contains many `ExecStartPre`/`ExecStartPost` commands, with
+  little explanation or infrastructure.  One `ExecStopPost` command depends on
+  `/bin/sh`, which is not available in Bottlerocket or compatible with our
+  Bottlerocket's goals.
```
rephrase: "our Bottlerocket's goals" -> "our goals for Bottlerocket"

samuelkarp replied:
...I thought I fixed that already.  Oops!

samuelkarp replied:
Done

samuelkarp committed:
.github: add aws-ecs-1 variant to Build workflow

bcressey commented on sources/models/src/lib.rs:
```diff
@@ -99,6 +106,14 @@ struct KubernetesSettings {
     pod_infra_container_image: SingleLineString,
 }
 
+// ECS settings.
+#[model]
+struct ECSSettings {
+    cluster: String,
+    instance_attributes: HashMap<ECSAttributeKey, ECSAttributeValue>,
+    privileged_containers_enabled: bool,
+}
+
```
Maybe "allow_privileged_containers" or "enable_privileged_containers". We don't have a lot of settings like this so it's hard to point to a convention, but `settings.updates.ignore-waves` seems closest.

samuelkarp replied:
I like "allow_privileged_containers"; will change to that.

samuelkarp replied:
Done

bcressey commented on packages/ecs-agent/ecs.service:
```diff
@@ -0,0 +1,34 @@
+EnvironmentFile=-/etc/ecs/ecs.config
+Environment=ECS_LOGFILE=/var/log/ecs/ecs-agent.log ECS_LOGLEVEL=debug
+Environment=ECS_CHECKPOINT=true
+ExecStartPre=mkdir -p /var/log/ecs /var/lib/ecs
```
I'd prefer to create these via a tmpfiles snippet; we do this for chrony and wicked. That would also simplify relabeling them if we need to.

samuelkarp replied:
I did not know about [tmpfiles.d](https://fedoraproject.org/wiki/Packaging:Tmpfiles.d) before.  That seems reasonable.

samuelkarp replied:
Done

bcressey commented on packages/ecs-agent/Cargo.toml:
```diff
@@ -0,0 +1,27 @@
+url = "https://github.com/aws/amazon-ecs-agent/archive/v1.41.0.tar.gz"
+sha512 = "d53d1a47eb41d39ffd541b6e416b42276f1d2207fdf5f031be173a8737655dab71ca0c88732716b6fee0761f7fede474cee08a21fd2db5d33b460d6aa547630e"
+
+# Pause container
+
+# awsvpc CNI plugin
+
+# Other CNI plugin?
+
+#[[package.metadata.build-package.external-files]]
+#url = "https://s3.amazonaws.com/amazon-ecs-agent/ecs-agent-v1.40.0.tar"
+#sha512 = "1f22be420dc42aeefa249ef56fe4a594159c97d65ff65ee777fa7823a539e35b917c40d7679cccf3401b90bba045b53a6f3b17c53a034392295c935cc1c59cfa"
+
```
Should we clean these up?

samuelkarp replied:
Yes, though I'm going to change it to be a more clear TODO comment.  The CNI plugins are included upstream via git submodules but not included in the tar, so they'll need to be brought in as other sources.  I'm not planning to actually *do* that as part of this PR, but I want to make sure there's a note marking the remaining work.

samuelkarp replied:
Done

bcressey commented on packages/ecs-agent/ecs-agent.spec:
```diff
@@ -0,0 +1,102 @@
+BuildRequires: %{_cross_os}glibc-devel
+
+Requires: %{_cross_os}docker-engine
+# for sysctl
+Requires: %{_cross_os}procps
```
We might not need this if we decide not to try to reset the `localnet` value in the service unit.

samuelkarp replied:
It's still used for `ExecStartPre=/sbin/sysctl -w net.ipv4.conf.all.route_localnet=1`.  Would you prefer that we directly write to the `/proc/sys/net/ipv4/conf/all/route_localnet` file instead?

bcressey replied:
We can use a sysctl.d setting like `release-sysctl.conf`, but included in the agent package.

samuelkarp replied:
Done

bcressey commented on packages/ecs-agent/ecs-agent.spec:
```diff
@@ -0,0 +1,102 @@
+  mkdir -p image/rootfs
+  %tar_cf image/rootfs/layer.tar -C rootfs .
+  DIGEST=$(sha256sum image/rootfs/layer.tar | sed -e 's/ .*//')
+  echo -n "1.0" > image/rootfs/VERSION
```
Should this be a source file also? That make it so all the contents of the container came from external files.

samuelkarp replied:
Done

bcressey commented on packages/ecs-agent/ecs-agent.spec:
```diff
@@ -0,0 +1,102 @@
+install -D -p -m 0755 amazon-ecs-agent %{buildroot}%{_cross_bindir}/amazon-ecs-agent
+install -D -p -m 0644 amazon-ecs-pause.tar %{buildroot}%{_cross_libdir}/amazon-ecs-agent/amazon-ecs-pause.tar
+
+mkdir -p %{buildroot}%{_cross_sysconfdir}/ecs
+touch %{buildroot}%{_cross_sysconfdir}/ecs/ecs.config.json
```
Why do this? `/etc` is covered by a tmpfs mount very early during boot, so any file there won't be visible.

samuelkarp replied:
Removed

samuelkarp commented:
(GitHub won't let me reply on this)

@bcressey wrote:

> Do we need to set `net.ipv4.conf.all.route_localnet` back to the default value on stop? Might be easiest to just remove it.

Probably not; this is one of the items in the README.md that I was looking for another way around.  I don't really see a use-case for un-setting these options when ecs.service stops, but I also don't want to re-run the setup commands every time the service starts.  My current idea (which I have not tried yet) is to move all these setup commands to a separate oneshot systemd service that the main ecs.service depends on.

webern commented on sources/api/ecs-settings-applier/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+publish = false
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
maybe remove the cargo new boilerplate.

webern commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,150 @@
+    let settings = schnauzer::get_settings(&args.socket_path).context(error::SettingsError)?;
+
+    debug!("settings = {:#?}", settings.settings);
+    let ecs = settings.settings.and_then(|s| s.ecs);
```
Maybe it can be ref-ified here to eliminate the subsequent .as_ref() calls? (Not sure).

```suggestion
    let ecs = settings.settings.and_then(|s| s.ecs).as_ref();
```

I was also trying to see if there was some simplified code path that could eliminate the profusion of `if_then`s, but I couldn't see it.

Sometimes you can do, like

```rust
let ecs = if let Some(x) = ecs {
    x
} else {
   // do something simple and bail
  return Ok(());
}

// lots of code where ecs is no longer an Option
```

Maybe this doesn't work for you though since os.settings is also an option.

webern replied:
P.S. - not a blocker for me.

samuelkarp replied:
@webern Thanks! Going to take a look and see if I can take your suggestions.  I tried a bunch of things with `as_ref()` and `if_then` when I was writing this, but I am very much a Rust newbie.

samuelkarp replied:
```
error[E0716]: temporary value dropped while borrowed
  --> api/ecs-settings-applier/src/main.rs:48:15
   |
48 |     let ecs = settings.settings.and_then(|s| s.ecs).as_ref();
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^         - temporary value is freed at the end of this statement
   |               |
   |               creates a temporary which is freed while still in use
49 |     let cluster = ecs.and_then(|s| s.cluster.as_ref());
   |                   --- borrow later used here
```

I was able to use your `let` suggestion though; see the latest push for a commit making that change.

webern reviewed (approved):
Impressive opus!

samuelkarp committed:
ecs-settings-applier: destructure ecs settings

samuelkarp committed:
ecs-agent: use systemd-sysctl and sysctl.d

etungsten commented on packages/ecs-agent/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+# the source of those plugins.
+
+[build-dependencies]
+glibc = { path = "../glibc" }
```
nit: Newline here would be nice

Same comment for `packages/ecs-agent/ecs-sysctl.conf` and `packages/ecs-agent/ecs-tmpfiles.conf`

samuelkarp replied:
Done

etungsten commented on packages/ecs-agent/ecs-agent.spec:
```diff
@@ -0,0 +1,103 @@
+
+# Upstream: https://github.com/aws/amazon-ecs-agent/pull/2513
+# Upstream status: Merged
+Patch0001: 0001-engine-move-default-image-exclusions.patch
```
If it's merged, does that mean this patch can be removed?

etungsten replied:
Oh but it's not yet released.

samuelkarp replied:
No, it's not yet present in a release: https://github.com/aws/amazon-ecs-agent/commit/9a03cfb3f732b753375f43afb815724f177d00bd.

etungsten commented on packages/os/os.spec:
```diff
@@ -183,6 +190,13 @@ for crate in $(find %{_builddir}/sources/api/migration/migrations -name 'Cargo.t
     %cargo_build_static --manifest-path "${crate}"
 done
 
+# Build conditional ECS component
```
nit: Maybe move this above the building of migrations?

samuelkarp replied:
Done

etungsten commented on sources/api/ecs-settings-applier/Cargo.toml:
```diff
@@ -0,0 +1,18 @@
+publish = false
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
```suggestion
```

samuelkarp replied:
Done

etungsten commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,158 @@
+    let cluster = ecs.cluster.as_ref();
+    let privileged_disabled = ecs.allow_privileged_containers.map(|s| !s);
+    let mut config = ECSConfig{
+        cluster: cluster.map(|s| s.clone()),
```
Try `.cloned()` here, and we can get rid of the intermediary `cluster`
```suggestion
        cluster: ecs.cluster.clone()
```

samuelkarp replied:
Done (by Jake's commit)

jahkeup committed:
Unpack expected Options

etungsten commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,158 @@
+            config.instance_attributes.insert(VERSION_ATTRIBUTE_NAME.to_string(), os.version_id.to_string());
+        }
+    }
+    match ecs.instance_attributes.as_ref() {
+        None => {}
+        Some(attributes) => {
+            for (key, value) in attributes {
+                config.instance_attributes.insert(key.to_string(), value.to_string());
+            }
+        }
+    }
```
I don't think `as_ref` is needed here:
```suggestion
    match ecs.instance_attributes {
        None => {}
        Some(attributes) => {
            for (key, value) in attributes {
                config.instance_attributes.insert(key.to_string(), value.to_string());
            }
        }
    }
```

Actually I think this and the match above (for `settings.os`) can be simplified to something like the following:
```suggestion
    if let Some(attributes) = ecs.instance_attributes {
        for (key, value) in attributes {
                config.instance_attributes.insert(key.to_string(), value.to_string());
        }
    }
```

samuelkarp replied:
Done (by Jake's commit)

etungsten commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,158 @@
+        fs::create_dir_all(dirname)?;
+    };
+
+    fs::write(path, contents).map(|_| ())
```
I think you can just do without the closure?

```suggestion
    fs::write(path, contents)
```

samuelkarp replied:
Done

etungsten commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,158 @@
+#[macro_use]
```
A simple README here would be nice for `ecs-settings-applier`

See the other crates in `/api` for examples

samuelkarp replied:
Done

jahkeup commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,158 @@
+
+    debug!("settings = {:#?}", settings.settings);
+    let ecs = settings.settings.and_then(|s| s.ecs);
+    let ecs = if let Some(x) = ecs {
```
I tried to see what other shapes this could take while adding some development changes:

https://github.com/jahkeup/bottlerocket/commit/15883b2eb3749d49d13446f36410629a39572068

There's a few other places in this code where the Option handling can be simplified because `None` isn't an error condition and is "Just data".

samuelkarp replied:
I took your commit; thanks!

etungsten commented on variants/README.md:
```diff
@@ -130,20 +135,6 @@ We use the same Rust code for all variants.
 
 ### Next Steps
 
-After the above files are in place, the variant must be added to the `variants` workspace.
```
Why did we remove this? Is this no longer applicable?

samuelkarp replied:
Correct, this is no longer applicable.

etungsten commented on sources/models/src/modeled_types/ecs.rs:
```diff
@@ -0,0 +1,172 @@
+            ECSAttributeValue::try_from(*err).unwrap_err();
+        }
+    }
+}
```
nit: newline here please!

samuelkarp replied:
Done

etungsten commented on sources/models/src/modeled_types/ecs.rs:
```diff
@@ -0,0 +1,172 @@
+
+    #[test]
+    fn bad_keys() {
+        for err in &[
```
`err` is a some what misleading name for this.
```suggestion
        for key in &[
```

samuelkarp replied:
These names were copied from here: https://github.com/bottlerocket-os/bottlerocket/blob/56bd7c22129854b56bec7b03e2ca90b16308cf07/sources/models/src/modeled_types/kubernetes.rs#L115-L139.  I'm happy to change them, but we might want to change them in the other file as well.

etungsten replied:
This is mostly me almost getting confused with the `ECSAttributeValue::try_from(*err)` below and for a second had flashbacks of golang. It's a non-blocker for me.

etungsten commented on sources/models/src/modeled_types/ecs.rs:
```diff
@@ -0,0 +1,172 @@
+
+    #[test]
+    fn good_keys() {
+        for ok in &[
```
nit:
```suggestion
        for key in &[
```

etungsten commented on sources/models/src/modeled_types/ecs.rs:
```diff
@@ -0,0 +1,172 @@
+
+    #[test]
+    fn bad_keys() {
+        for err in &[
```
`err` is a some what misleading name for this.
```suggestion
        for key in &[
```

etungsten commented on sources/models/src/modeled_types/ecs.rs:
```diff
@@ -0,0 +1,172 @@
+
+    #[test]
+    fn good_keys() {
+        for ok in &[
```
nit: 
```suggestion
        for key in &[
```

bcressey commented on packages/ecs-agent/ecs.service:
```diff
@@ -0,0 +1,30 @@
+[Unit]
+Description=Amazon Elastic Container Service - container agent
+Documentation=https://aws.amazon.com/documentation/ecs/
+Requires=docker.service
+After=docker.service configured.target
```
Changes to consider:
```
Requires=docker.service
BindsTo=docker.service
PartOf=docker.service
After=network-online.target docker.service configured.target
Wants=network-online.target configured.target
```

Rationale for BindsTo + After for `docker.service`:
> When used in conjunction with After= on the same unit the behaviour of BindsTo= is even stronger. In this case, the unit bound to strictly has to be in active state for this unit to also be in active state. This not only means a unit bound to another unit that suddenly enters inactive state, but also one that is bound to another unit that gets skipped due to a failed condition check... Hence, in many cases it is best to combine BindsTo= with After=.

Rationale for Wants - it's what we do elsewhere.

Rationale for PartOf:
> Configures dependencies similar to Requires=, but limited to stopping and restarting of units. When systemd stops or restarts the units listed here, the action is propagated to this unit.

We haven't used `PartOf` but it might be needed so that we also restart `ecs` when `docker` is restarted. That behavior may be useful if we need to reconfigure and restart docker in response to an API settings change.

`BindTo` and `PartOf` may not be needed if `ecs` can gracefully handle `docker` going away and coming back.

Test cases:
* when docker and ecs are stopped, `systemctl start ecs` should start docker
* when docker and ecs are running, `systemctl stop docker` should stop ecs
* when docker and ecs are running, `systemctl restart docker` should restart ecs
* when docker is running and ecs is stopped, `systemctl restart docker` should restart ecs (maybe?)

samuelkarp replied:
> `BindTo` and `PartOf` may not be needed if `ecs` can gracefully handle `docker` going away and coming back.

The ECS agent can gracefully handle Docker going away and coming back.  I don't see a need to restart the ECS agent when Docker is restarted.

I can definitely add `Wants` though.

samuelkarp replied:
Done

samuelkarp committed:
ecs-settings-applier: add README with cargo-readme

etungsten commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,163 @@
+        .context(error::ModelError)?;
+
+    let mut config = ECSConfig {
+        cluster: ecs.cluster.map(|s| s.to_owned()),
```
I don't think you need the `.map()` or even `.clone()` here.
Can you try this and see if it works?
```suggestion
        cluster: ecs.cluster,
```

samuelkarp replied:
Done, thanks!

samuelkarp committed:
ecs-settings-applier: simplify ECSConfig cluster

tjkirch commented on packages/ecs-agent/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+
+# ECS agent
+[[package.metadata.build-package.external-files]]
+url = "https://github.com/aws/amazon-ecs-agent/archive/v1.41.0.tar.gz"
```
1.41.1?

samuelkarp replied:
I'm expecting that there will be a newer version than 1.41.1 before the variant is really ready, so unless you think it's blocking to have an old version right now I'd prefer to handle the bump in a subsequent PR.  (1.41.1 doesn't have my patch in it yet, but the next release will, at which point I can drop the patch from the spec file as well.)

tjkirch commented on packages/ecs-agent/ecs-agent.spec:
```diff
@@ -0,0 +1,103 @@
+%global gitrev 3776bee9
+
+%global source_date_epoch 1234567890
+%global tar_cf tar --sort=name --mtime="@%{source_date_epoch}" --owner=0 --group=0 --numeric-owner -cf
```
A comment about reproducible tarballs might help explain this.

samuelkarp replied:
Done

tjkirch commented on sources/api/ecs-settings-applier/README.tpl:
```diff
@@ -0,0 +1,9 @@
+
+## Colophon
+
+This text was generated using [cargo-readme](https://crates.io/crates/cargo-readme), and includes the rustdoc from `src/main.rs`.
```
```suggestion
This text was generated from `README.tpl` using [cargo-readme](https://crates.io/crates/cargo-readme), and includes the rustdoc from `src/main.rs`.
```

nit: I prefer the newer form that says what it was generated from.

samuelkarp replied:
Done

tjkirch commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,163 @@
+# Introduction
+
+ecs-settings-applier generates a configuration file for the ECS agent from Bottlerocket settings.
+
```
```suggestion
```
nit: it's adding an extra line to the generated file.

samuelkarp replied:
Done

tjkirch commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,163 @@
+/*!
+# Introduction
+
+ecs-settings-applier generates a configuration file for the ECS agent from Bottlerocket settings.
```
I think it'd help to explain here why we need a binary for this and not just templates, as we found out.

samuelkarp replied:
Done

tjkirch commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,163 @@
+*/
+
+#[macro_use]
+extern crate log;
```
```suggestion
```
And then below, since I think it's only using `debug!`:
```
use log::debug;
```
(This is the newer form)

samuelkarp replied:
Done

tjkirch commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,163 @@
+        }
+    }
+
+    let serialized = serde_json::to_string(&config).unwrap();
```
Can we use a serialization error variant instead of unwrapping?

samuelkarp replied:
Done

tjkirch commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,163 @@
+    let serialized = serde_json::to_string(&config).unwrap();
+    debug!("serialized = {}", serialized);
+
+    let config_path = PathBuf::from(DEFAULT_ECS_CONFIG_PATH);
+    write_to_disk(config_path, serialized).context(error::FSError {
```
```suggestion
    write_to_disk(DEFAULT_ECS_CONFIG_PATH, serialized).context(error::FSError {
```
`&str` is `AsRef<Path>` :)

samuelkarp replied:
...I don't know why I did that.

samuelkarp replied:
Done

tjkirch commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,163 @@
+#[macro_use]
+extern crate log;
+
+use error::SettingsApplierError;
```
```suggestion
```
nit: a bit much, for a module in the same file.

samuelkarp replied:
Done

tjkirch commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,163 @@
+    process::exit(2);
+}
+
+type Result<T> = std::result::Result<T, SettingsApplierError>;
```
```suggestion
type Result<T> = std::result::Result<T, error::Error>;
```
Per above and below suggestions.

samuelkarp replied:
Done

tjkirch commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,163 @@
+
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility = "pub(super)")]
+    pub(super) enum SettingsApplierError {
```
```suggestion
    pub(super) enum Error {
```
It's already scoped to this module, don't need to repeat the name :)

samuelkarp replied:
Done

tjkirch commented on sources/models/src/modeled_types/ecs.rs:
```diff
@@ -0,0 +1,172 @@
+// underscores, and periods are allowed.
+lazy_static! {
+    pub(crate) static ref ECS_ATTRIBUTE_KEY: Regex = Regex::new(
+        r"(?x)^
+       [a-zA-Z0-9-_\./]{1,128}
+   $"
```
```suggestion
        r"(?x)^
          [a-zA-Z0-9_./-]{1,128}
          $"
```
* spacing suggestion
* you don't need to escape `.` in `[]`
* it's convention to put `-` last so someone doesn't think it's another range.

samuelkarp replied:
> you don't need to escape `.` in `[]`

`.` isn't escaped, a `\` is a valid character in the attribute name.  Now that the documentation is updated, I'll update the comment above this.

Rewriting the character set to `[a-zA-Z0-9._/\-]`.

tjkirch replied:
`\` is a valid character, but you have to escape it, right?  So `\\`?  With `\-` you're escaping `-` which doesn't need escaping and so is the same as just writing `-`.

samuelkarp replied:
Yeah, I had to change it to `[a-zA-Z0-9._/\\-]`; I wrote the previous comment before actually testing it.

samuelkarp replied:
The bigger issue is that the datastore is preventing me from having a key with a backslash:

```
$ apiclient -u /settings -m PATCH -d '{"ecs":{"instance-attributes":{"foo\\/":"\\ /bar \\"}}}'
Status 500 when PATCHing unix://2f72756e2f6170692e736f636b:0/settings: Error serializing Settings: Error creating valid datastore key: Key name 'foo\/' has invalid format: Segment 'foo\/' contains invalid character '\' 
```

tjkirch replied:
I vote for banning evil backslashes.

samuelkarp replied:
I'll remove them and add this to the README in ecs-agent where I'm tracking work.

samuelkarp replied:
Done

tjkirch commented on sources/models/src/modeled_types/ecs.rs:
```diff
@@ -0,0 +1,172 @@
+// underscores, periods, at signs (@), forward slashes, colons, and spaces are allowed.
+lazy_static! {
+    pub(crate) static ref ECS_ATTRIBUTE_VALUE: Regex = Regex::new(
+        r"(?x)^
+       [a-zA-Z0-9-_\\.@/:] # at least one non-space
+       (
+         (?-x)([a-zA-Z0-9-_\\.@/: ]{0,126})?(?x) # spaces allowed
+         [a-zA-Z0-9-_\\.@/:] # end with non-space
+       )?
+   $"
```
```suggestion
        r"(?x)^
          [a-zA-Z0-9_.@/:-] # at least one non-space
          (
            ([a-zA-Z0-9_.@/:\ -]{0,126})? # spaces allowed
            [a-zA-Z0-9_.@/:-] # end with non-space
          )?
          $"
```
Same suggestions as above, plus:
* this one had `\\.` instead of `\.` and I'm not sure if you intended it, but since you didn't mention backslashes in the comment and you don't need to escape `.` inside `[]`, I removed it in the suggestion...
* don't need to escape and reenter `x` mode, can just escape a space inside.

samuelkarp replied:
The [documentation](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_Attribute.html) has been updated since I wrote the comment; backslashes are allowed in attribute values.

I'll try the other suggestions, thanks!

samuelkarp replied:
Done

tjkirch reviewed (approved):
I have to trust you on a bunch of this, but I can't think of anyone better to trust :)

iliana commented on packages/os/os.spec:
```diff
@@ -147,6 +147,13 @@ Summary: Thar data store migrations
 %description -n %{_cross_os}migrations
 %{summary}.
 
+%if "%{_cross_variant}" == "aws-ecs-1"
+%package -n %{_cross_os}ecs-settings-applier
```
Can this only be built on the aws-ecs-1 variant due to a dependency on ECS-specific models?

samuelkarp replied:
Correct

tjkirch commented on sources/api/ecs-settings-applier/src/main.rs:
```diff
@@ -0,0 +1,166 @@
+        ModelError,
+
+        #[snafu(display("Failed to serialize ECS config: {}", source))]
+        SerializationError {
```
```suggestion
        Serialization {
```
nit: since you always reference it as `error::Bla`, it's clear that it's an error.  (Same for the others)

samuelkarp replied:
Done

samuelkarp committed:
ecs-settings-applier: cargo fmt

fenxiong commented on packages/ecs-agent/0002-bottlerocket-default-filesystem-locations.patch:
```diff
@@ -0,0 +1,82 @@
+ 		ReservedPorts:                       []uint16{SSHPort, DockerReservedPort, DockerReservedSSLPort, AgentIntrospectionPort, AgentCredentialsPort},
+ 		ReservedPortsUDP:                    []uint16{},
+-		DataDir:                             "/data/",
++		DataDir:                             "/var/lib/ecs",
```
the path `/data/` was normally to map to path `/var/lib/ecs/data` on the host when agent is running in container, so this actually changes the data dir on the host from `/var/lib/ecs/data` to `/var/lib/ecs`. is it intentional? some feature, e.g. firelens, is assuming that the files are stored in the data folder (e.g. [here](https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/task/task.go#L107)) so changing it here will force the path to diverge in the feature code path as well

samuelkarp replied:
I can change it to `/var/lib/ecs/data` if desired.  However, it seems like a bug that the Firelens code will assume that to be the path rather than reading the path from the config.

fenxiong replied:
Well, the agent already reads the host path from config.DataDirOnHost, which does not contain the "data" folder so we need to append "data/" to it. The complication is caused by the fact that config.DataDirOnHost does not contain folder "data" but config.DataDir does.

samuelkarp replied:
Yeah, the current code is coupled to that assumption.  To me that seems like a bug in the design; I'll see if I can come up with some suggestions for making it nicer once I get to the bits that need it.

Right now, with `config.DataDirOnHost` set to `/var/lib/ecs`, the Firelens bind ends up being `/var/lib/ecs/data/firelens/%s/config/fluent.conf:/fluentd/etc/fluent.conf`, which seems reasonable.  I don't think it's accessing `ecs_agent_state.json`, so it doesn't matter whether it's in the same folder (`/var/lib/ecs`) or not (`/var/lib/ecs/data`).

samuelkarp replied:
Done

fenxiong commented on packages/ecs-agent/README.md:
```diff
@@ -0,0 +1,23 @@
+This README is temporary and is meant to track the known issues and remaining
+work items for the ECS agent.
+
+## Known issues
```
nit: two other things that i can see that remain to be addressed and not mentioned here:
1. log path of cni plugin, currently all of them are defaulting to the container path of the log directory bind mount,  e.g.  [here](https://github.com/aws/amazon-ecs-cni-plugins/blob/0c6216c60401232805e50e31d4040ae84b0b23cf/plugins/eni/main.go#L32), [here](https://github.com/aws/amazon-ecs-agent/blob/master/agent/ecscni/plugin.go#L39)
2. path to process's network namespace handle currently hardcoded to the container path of the procfs bind mount https://github.com/aws/amazon-ecs-agent/blob/782948476da6d995ad33c6a04130f8172820af27/agent/ecscni/types.go#L38

would be good to track them

samuelkarp replied:
Sure, I can add those.  However, none of the CNI-related features are currently supported as is called out in the second bullet point.

samuelkarp replied:
Done
